File: FrankGUI/SanityChecker.m
//
//  SanityChecker.m
//  FrankGUI
//
//  Created by Oleksiy Radyvanyuk on 10/02/15.
//  Copyright (c) 2015 Oleksiy Radyvanyuk. All rights reserved.
//

#import "SanityChecker.h"
#import "ConsoleToolExecutor.h"
#import "Constants.h"
#import "Settings.h"

@interface SanityChecker ()

@property (nonatomic, strong) Settings *settings;
@property (nonatomic, strong) ConsoleToolExecutor *executor;
@property (nonatomic, strong) NSString *gitLaunchPath;
@property (nonatomic, strong) NSString *gitBranchNameInAppPathURL;
@property (nonatomic, strong) NSString *gitBranchNameInScriptsPathURL;
@property (nonatomic, strong) NSString *gemLaunchPath;
@property (nonatomic, strong) NSString *frankCucumberGemVersion;
@property (nonatomic, strong) NSString *frankCucumberGemDirectoryName;
@property (nonatomic, strong) NSString *frankCucumberGemPath;

- (void)setupToolsPaths;
- (NSString *)gitBranchNameInDirectory:(NSString *)directory;
- (BOOL)isValidRepositoryPathURL:(NSURL *)pathURL;

- (BOOL)isValidAppPathURL;
- (BOOL)isValidScriptsPathURL;
- (BOOL)areTheSameBranchesInAppAndInScriptsPaths;
- (BOOL)isValidFrankCucumberGemVersion;

@end

@implementation SanityChecker

- (instancetype)initWithSettings:(Settings *)settings
{
    if (self = [super init])
    {
        self.settings = settings;
        self.executor = [ConsoleToolExecutor new];
        [self setupToolsPaths];
    }

    return self;
}

- (void)setupToolsPaths
{
    // determine path to git tool
    self.gitLaunchPath = [self.executor pathForTool:@"git"];

    // determine path to gem tool
    self.gemLaunchPath = [self.executor pathForTool:@"gem"];
}

- (NSString *)gitBranchNameInDirectory:(NSString *)directory
{
    NSArray* args = @[@"status", @"-s", @"-b", @"--porcelain"];
    int exitCode = -1;
    NSString *output = [self.executor outputOfCommand:self.gitLaunchPath inDirectory:directory withArguments:args exitCode:&exitCode];

    if (0 == exitCode && [output length] > 0)
    {
        // extract branch name that is usually in format like this:
        //
        // ## master...origin/master
        // ?? External/MKMapViewZoom/
        
        // first leave the first line of the output only
        NSArray *lines = [output componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
        if ([lines count] < 1)
        {
            return nil;
        }

        NSArray *parts = [[lines firstObject] componentsSeparatedByString:@"..."];
        if ([parts count] > 0)
        {
            NSString *firstPart = [parts firstObject];
            NSArray *subParts = [firstPart componentsSeparatedByString:@" "];
            if ([subParts count] > 1)
            {
                return [subParts objectAtIndex:1];
            }
        }
    }

    return nil;
}

- (BOOL)isValidRepositoryPathURL:(NSURL *)pathURL
{
    // check that the given path points to an existing directory
    if ([pathURL isFileURL])
    {
        NSString *path = [pathURL path];
        BOOL pathIsDirectory = NO;
        BOOL pathExists = [[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&pathIsDirectory];
        
        return pathExists && pathIsDirectory;
    }

    return NO;
}

- (BOOL)isValidAppPathURL
{
    // there should be git branch available in that directory
    return [self isValidRepositoryPathURL:self.settings.appPathURL] && [self.gitBranchNameInAppPathURL length] > 0;
}

- (BOOL)isValidScriptsPathURL
{
    // there should be git branch available in that directory
    return [self isValidRepositoryPathURL:self.settings.scriptsPathURL] && [self.gitBranchNameInScriptsPathURL length] > 0;
}

- (BOOL)areTheSameBranchesInAppAndInScriptsPaths
{
    // both branches for app and for scripts must be the same
    return [self.gitBranchNameInAppPathURL isEqualToString:self.gitBranchNameInScriptsPathURL];
}

- (NSString *)frankCucumberGemName
{
    return @"ngti-frank-cucumber";
}

- (NSString *)frankCucumberGemVersion
{
    if (nil == _frankCucumberGemVersion)
    {
        NSArray *args = @[@"list", @"-l", self.frankCucumberGemName];
        int exitCode = -1;
        NSString *output = [self.executor outputOfCommand:self.gemLaunchPath inDirectory:nil withArguments:args exitCode:&exitCode];

        if (0 == exitCode && [output length] > 0)
        {
            _frankCucumberGemVersion = output;
        }
        else
        {
            _frankCucumberGemVersion = nil;
        }
    }

    return _frankCucumberGemVersion;
}

- (NSString *)frankCucumberGemDirectoryName
{
    if (nil == _frankCucumberGemDirectoryName)
    {
        NSString *gemString = self.frankCucumberGemVersion;
        if (nil != gemString)
        {
            NSArray *gemStringParts = [gemString componentsSeparatedByCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@"()"]];

            if ([gemStringParts count] > 1)
            {
                _frankCucumberGemDirectoryName = [NSString stringWithFormat:@"%@-%@", self.frankCucumberGemName, gemStringParts[1]];
            }
        }
    }

    return _frankCucumberGemDirectoryName;
}

- (NSString *)frankCucumberGemPath
{
    if (nil == _frankCucumberGemPath)
    {
        // obtain search paths of gems
        NSArray *arguments = @[@"environment", @"gempath"];
        int exitCode = -1;
        NSString *output = [self.executor outputOfCommand:self.gemLaunchPath inDirectory:nil withArguments:arguments exitCode:&exitCode];

        // only continue with successful output and exit code
        if (nil != output && 0 == exitCode)
        {
            NSArray *paths = [output componentsSeparatedByString:@":"];
            // begin searching through paths for the paritcular directory
            NSString *gemDirName = self.frankCucumberGemDirectoryName;
            NSFileManager *fileMan = [NSFileManager defaultManager];
            for (NSString *path in paths)
            {
                // construct candidate directory of gems common base directory and particular gem name
                NSString *gemDir = [NSString stringWithFormat:@"%@/gems/%@", path, gemDirName];
                BOOL isDirectory = NO;
                if ([fileMan fileExistsAtPath:gemDir isDirectory:&isDirectory])
                {
                    if (isDirectory)
                    {
                        _frankCucumberGemPath = gemDir;
                        break;
                    }
                }
            }
        }
    }

    return _frankCucumberGemPath;
}

- (BOOL)isValidFrankCucumberGemVersion
{
    // get the full name and version of the gem
    NSString *gemString = self.frankCucumberGemVersion;
    NSArray *gemStringParts = [gemString componentsSeparatedByCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@"()"]];

    if (3 != [gemStringParts count])
    {
        // expect gem string in format <name> (<version>)
        return NO;
    }
    
    NSArray *versionParts = [gemStringParts[1] componentsSeparatedByString:@"."];
    if (3 != [versionParts count])
    {
        // expect version components like <major>.<minor>.<build>
        return NO;
    }
    
    // expect versions of 1.3.x
    if ([versionParts[0] intValue] < 1 || [versionParts[1] intValue] < 3)
    {
        return NO;
    }

    return YES;
}

- (void)validate
{
    [self.delegate validatingDidStart];
    self.gitBranchNameInAppPathURL = [self gitBranchNameInDirectory:[self.settings.appPathURL path]];
    self.gitBranchNameInScriptsPathURL = [self gitBranchNameInDirectory:[self.settings.scriptsPathURL path]];

    if (![self isValidAppPathURL])
    {
        [self.delegate validatingWarnLevel:WarnLevelError message:@"Error! Incorrect path to app sources"];
        [self.delegate validatingDidFinish];
        return;
    }

    if (![self isValidScriptsPathURL])
    {
        [self.delegate validatingWarnLevel:WarnLevelError message:@"Error! Incorrect path to Frank scripts"];
        [self.delegate validatingDidFinish];
        return;
    }

    // verifying correctness of "ngti-frank-cucumber" gem version
    NSString *gemVersion = self.frankCucumberGemVersion;
    if (nil == gemVersion || ![self isValidFrankCucumberGemVersion])
    {
        NSString *message = [NSString stringWithFormat:@"Error! Cannot determine version of gem '%@'! Make sure it is installed.", self.frankCucumberGemName];
        [self.delegate validatingWarnLevel:WarnLevelError message:message];
        [self.delegate validatingDidFinish];
        return;
    }

    // verifying correctness of both branches
    if (![self areTheSameBranchesInAppAndInScriptsPaths])
    {
        [self.delegate validatingWarnLevel:WarnLevelIssue message:@"Warning! App has checked out of branch different from Frank scripts branch.\nMake sure you understand what you are doing!"];
        [self.delegate validatingDidFinish];
        return;
    }

    // if we got here, all the checks have passed
    [self.delegate validatingWarnLevel:WarnLevelOK message:@"Ready to run!"];
    [self.delegate validatingDidFinish];
}

- (NSArray *)listOfAvailablePlatforms
{
    NSString *envDir = [[self.settings.scriptsPathURL relativePath] stringByAppendingPathComponent:kFrankSkeletonEnv];

    NSMutableArray *result = [NSMutableArray new];

    NSDirectoryEnumerator *enumerator = [[NSFileManager defaultManager] enumeratorAtPath:envDir];
    for (NSString *file in enumerator)
    {
        // only include those ending with .sh
        if ([[file pathExtension] isEqualToString:@"sh"])
        {
            // cut off the extension, turn base file name into upper case
            NSString *baseName = [[file stringByDeletingPathExtension] uppercaseString];
            [result addObject:baseName];
        }
    }

    return result;
}

@end

File: FrankGUI/SanityChecker.swift
//
//  SanityChecker.swift
//  FrankGUI
//
//  Created by Oleksiy Radyvanyuk on 16/06/15.
//  Copyright © 2015 Oleksiy Radyvanyuk. All rights reserved.
//

import Foundation

@objc
protocol SanityCheckerDelegate
{
    /**
    * Notifies delegate that validating did start.
    */
    func validatingDidStart()

    /**
    * Notifies delegate that validating did finish. All the warn levels of validating are reported
    * via the method -validatingWarnLevel:message:
    */
    func validatingDidFinish()

    /**
    * Notifies delegate that certain warn level was determined.
    *
    * @param warnLevel Warn level determined
    * @param message A message that corresponds to the warn level
    */
    func validatingWarnLevel(warnLevel: SanityChecker.WarnLevel, message: String)
}

final class SanityChecker : NSObject
{
    /**
    * Result of the last sanity check as a whole. The value dtermines if this is safe
    * to operate in the app. All levels except for WarnLevelError are considered safe
    * to operate.
    */
    @objc enum WarnLevel: Int
    {
        case OK    // all sanity checks have passed successfully
        case Issue // there were some issues discovered during sanity checks
        case Error // critical failure has occured while validating
    }

    var delegate: SanityCheckerDelegate?

    private(set) var gitBranchNameInAppPathURL: String?
    private(set) var gitBranchNameInScriptsPathURL: String?
    var frankCucumberGemName: String {
        get {
            return "ngti-frank-cucumber"
        }
    }

    private var _frankCucumberGemVersion: String?
    var frankCucumberGemVersion: String? {
        get {
            if (_frankCucumberGemVersion == nil)
            {
                let args = ["list", "-l", frankCucumberGemName]
                var exitCode: Int32 = -1
                var output: String? = nil
                executor.outputOfCommand(gemLaunchPath,
                    inDirectory: nil,
                    withArguments: args,
                    output:
                    { (internalResult: String, internalExitCode: Int32) in
                        output = internalResult
                        exitCode = internalExitCode
                    }
                )
                
                if ((exitCode == 0) && (output?.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) > 0))
                {
                    _frankCucumberGemVersion = output
                }
                else
                {
                    _frankCucumberGemVersion = nil
                }
            }
            return _frankCucumberGemVersion
        }
    }

    private var _frankCucumberGemDirectoryName: String?
    private var frankCucumberGemDirectoryName: String? {
        get {
            if (_frankCucumberGemDirectoryName == nil)
            {
                if let gemString = frankCucumberGemVersion
                {
                    let gemStringParts = gemString.componentsSeparatedByCharactersInSet(NSCharacterSet(charactersInString: "()"))

                    if (gemStringParts.count > 1)
                    {
                        _frankCucumberGemDirectoryName = "\(frankCucumberGemName)-\(gemStringParts[1])"
                    }
                }
            }
            return _frankCucumberGemDirectoryName
        }
    }
    
    private var _frankCucumberGemPath: String?
    var frankCucumberGemPath: String? {
        get {
            if (_frankCucumberGemPath == nil)
            {
                // obtain search paths of gems
                let arguments = ["environment", "gempath"]
                var exitCode: Int32 = -1
                var output: String? = nil
                executor.outputOfCommand(gemLaunchPath,
                    inDirectory: nil,
                    withArguments: arguments,
                    output:
                    { (internalResult: String, internalExitCode: Int32) in
                        output = internalResult
                        exitCode = internalExitCode
                    }
                )

                // only continue with successful output and exit code
                if ((exitCode == 0) && (output?.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) > 0))
                {
                    let paths = output!.componentsSeparatedByString(":")
                    // begin searching through paths for the paritcular directory
                    let gemDirName = frankCucumberGemDirectoryName
                    let fileMan: NSFileManager = NSFileManager.defaultManager()
                    for path: String in paths
                    {
                        // construct candidate directory of gems common base directory and particular gem name
                        let gemDir = "\(path)/gems/\(gemDirName)"
                        var isDirectory: ObjCBool = false
                        if (fileMan.fileExistsAtPath(gemDir, isDirectory:&isDirectory))
                        {
                            if (isDirectory)
                            {
                                _frankCucumberGemPath = gemDir
                                break
                            }
                        }
                    }
                }
            }
            return _frankCucumberGemPath
        }
    }

    private var settings: Settings
    private var executor: ConsoleToolExecutor

    init(settings: Settings)
    {
        self.settings = settings
        executor = ConsoleToolExecutor()

        gitBranchNameInAppPathURL = ""
        gitBranchNameInScriptsPathURL = ""

        super.init()
        setupToolsPaths()
    }

    func validate()
    {
        delegate?.validatingDidStart()
        gitBranchNameInAppPathURL = gitBranchNameInDirectory(settings.appPathURL?.path)
        gitBranchNameInScriptsPathURL = gitBranchNameInDirectory(settings.scriptsPathURL?.path)

        if (!isValidAppPathURL())
        {
            delegate?.validatingWarnLevel(.Error, message: "Error! Incorrect path to app sources")
            delegate?.validatingDidFinish()
            return
        }

        if (!isValidScriptsPathURL())
        {
            delegate?.validatingWarnLevel(.Error, message: "Error! Incorrect path to Frank scripts")
            delegate?.validatingDidFinish()
            return
        }

        // verifying correctness of "ngti-frank-cucumber" gem version
        if ((frankCucumberGemVersion != nil) && !isValidFrankCucumberGemVersion())
        {
            let message = "Error! Cannot determine version of gem '\(frankCucumberGemName)'! Make sure it is installed."
            delegate?.validatingWarnLevel(.Error, message: message)
            delegate?.validatingDidFinish()
            return
        }

        // verifying correctness of both branches
        if (!areTheSameBranchesInAppAndInScriptsPaths())
        {
            delegate?.validatingWarnLevel(.Issue, message: "Warning! App has checked out of branch different from Frank scripts branch.\nMake sure you understand what you are doing!")
            delegate?.validatingDidFinish()
            return
        }

        // if we got here, all the checks have passed
        delegate?.validatingWarnLevel(.OK, message: "Ready to run!")
        delegate?.validatingDidFinish()
    }

    func listOfAvailablePlatforms() -> [String]
    {
        guard let envDir = settings.scriptsPathURL?.relativePath?.stringByAppendingPathComponent("env") else {
            // something went wrong - no path to "env" subdirectory
            return []
        }

        var result = [String]()

        let enumerator = NSFileManager.defaultManager().enumeratorAtPath(envDir)
        while let file = enumerator?.nextObject() as? String
        {
            // only include those ending with .sh
            if (file.pathExtension == "sh")
            {
                // cut off the extension, turn base file name into upper case
                let baseName = file.stringByDeletingPathExtension.uppercaseString
                result.append(baseName)
            }
        }

        return result
    }

    private var gitLaunchPath: String?
    private var gemLaunchPath: String?

    private func setupToolsPaths()
    {
        // determine path to git tool
        gitLaunchPath = executor.pathForTool("git")

        // determine path to gem tool
        gemLaunchPath = executor.pathForTool("gem")
    }

    private func gitBranchNameInDirectory(directory: String?) -> String?
    {
        let args = ["status", "-s", "-b", "--porcelain"]

        var exitCode: Int32 = -1
        var output: String = ""

        executor.outputOfCommand(gitLaunchPath,
            inDirectory: directory,
            withArguments: args,
            output:
            { (internalResult: String, internalExitCode: Int32) in
                output = internalResult
                exitCode = internalExitCode
            }
        )

        if (exitCode == 0) && (output.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) > 0)
        {
            // extract branch name that is usually in format like this:
            //
            // ## master...origin/master
            // ?? External/MKMapViewZoom/
            
            // first leave the first line of the output only
            let lines = output.componentsSeparatedByCharactersInSet(NSCharacterSet.newlineCharacterSet())
            if (lines.count < 1)
            {
                return nil
            }
            
            let parts = lines.first?.componentsSeparatedByString("...")
            if (parts?.count > 0)
            {
                let firstPart = parts?.first
                let subParts = firstPart?.componentsSeparatedByString(" ")
                
                if (subParts?.count > 1)
                {
                    return subParts?[1]
                }
            }
        }

        return nil
    }
    
    private func isValidRepositoryPathURL(pathURL: NSURL) -> Bool
    {
        // check that the given path points to an existing directory
        if (pathURL.fileURL)
        {
            let path = pathURL.path
            var pathIsDirectory: ObjCBool = false
            let pathExists = NSFileManager.defaultManager().fileExistsAtPath(path!, isDirectory: &pathIsDirectory)
            return pathExists && pathIsDirectory
        }

        return false
    }
    
    private func isValidAppPathURL() -> Bool
    {
        // there should be git branch available in that directory
        return isValidRepositoryPathURL(settings.appPathURL!) &&
            (gitBranchNameInAppPathURL!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) > 0)
    }

    private func isValidScriptsPathURL() -> Bool
    {
        // there should be git branch available in that directory
        return isValidRepositoryPathURL(settings.scriptsPathURL!) &&
            (gitBranchNameInScriptsPathURL!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) > 0)
    }
    
    private func areTheSameBranchesInAppAndInScriptsPaths() -> Bool
    {
        // both branches for app and for scripts must be the same
        return gitBranchNameInAppPathURL == gitBranchNameInScriptsPathURL
    }

    private func isValidFrankCucumberGemVersion() -> Bool
    {
        // get the full name and version of the gem
        let gemString = frankCucumberGemVersion
        let gemStringParts = gemString!.componentsSeparatedByCharactersInSet(NSCharacterSet(charactersInString: "()"))

        if (gemStringParts.count != 3)
        {
            // expect gem string in format <name> (<version>)
            return false
        }

        let versionParts = gemStringParts[1].componentsSeparatedByString(".")
        if (versionParts.count != 3)
        {
            // expect version components like <major>.<minor>.<build>
            return false
        }

        // expect versions of 1.3.x
        let versionPartMajor = Int(versionParts[0])
        let versionPartMinor = Int(versionParts[1])
        if (versionPartMajor < 1) || (versionPartMinor < 3)
        {
            return false
        }

        return true
    }
}
